File: 01-data-structures/03-linked-lists-js/linked_list.js
'use strict';

class LinkedList {
  constructor(node) {
    this.head = node || null;
    this.tail = node || null;
  }

  addToTail(node) {
    if (this.tail) {
      this.tail.next = node;
    } else {
      this.head = node;
    }

    this.tail = node;
  }

  removeTail() {
    if (!this.head) {
      return;
    } else if (this.head === this.tail) {
      this.head = null;
      this.tail = null;
    } else {
      let node = this.head;

      while (node) {
        if (node.next === this.tail) {
          this.tail = node;
          node.next = null;
          return;
        }
        node = node.next;
      }
    }
  }

  print() {
    let node = this.head;

    while (node) {
      console.log(`${node.data}\n`);
      node = node.next;
    }
  }

  remove(node) {
    if (!this.head) {
      return;
    } else if (this.head === node) {
      if (this.head === this.tail) {
        this.head = null;
        this.tail = null;
      } else {
        this.head = this.head.next;
      }
      return;
    } else {
      let tempNode = this.head;

      while (tempNode) {
        if (tempNode.next === node) {
          let toBeRemoved = tempNode.next;
          tempNode.next = tempNode.next.next;

          if (this.tail === toBeRemoved) {
            this.tail = tempNode;
          }

          toBeRemoved.next = null;
          return;
        }
        tempNode = tempNode.next;
      }
    }
  }

  addToFront(node) {
    if (!this.head) {
      this.head = node;
      this.tail = node;
      return;
    }

    const temp = this.head;

    this.head = node;
    this.head.next = temp;
  }

  removeFront() {
    if (!this.head) { return; }

    let temp = this.head;

    this.tail = this.head === this.tail ? null : this.tail;
    this.head = this.head.next;
    temp = null;
  }
}

module.exports = LinkedList;

File: 01-data-structures/05-hashes-part-2/separate_chaining/linked_list.rb
require_relative 'node'

class LinkedList
  attr_accessor :head
  attr_accessor :tail

  # This method creates a new `Node` using `data`, and inserts it at the end of the list.
  def add_to_tail(node)
    if @tail
      @tail.next = node
    else
      @head = node
    end

    @tail = node
  end

  # This method removes the last node in the lists and must keep the rest of the list intact.
  def remove_tail
    if !@head
      return
    elsif @head == @tail
      @head = nil
      @tail = nil
    else
      node = @head

      while node do
        if node.next == @tail
          @tail = node
          node.next = nil
          break
        end

        node = node.next
      end
    end
  end

  # This method prints out a representation of the list.
  def print
    node = @head

    while node do
      puts node.data
      node = node.next
    end
  end

  # This method removes `node` from the list and must keep the rest of the list intact.
  def delete(node)
    if !@head
      return
    elsif @head == node
      if @head == @tail
        @head == nil
        @tail == nil
      else
        @head = @head.next
      end
    else
      temp_node = @head

      while temp_node do
        if temp_node.next == node
          to_be_removed = temp_node.next
          temp_node.next = temp_node.next.next

          if @tail == to_be_removed
            @tail = temp_node
          end

          to_be_removed.next = nil
          break;
        end
        temp_node = temp_node.next
      end
    end
  end

  # This method adds `node` to the front of the list and must set the list's head to `node`.
  def add_to_front(node)
    if !@head
      @head = node
      @tail = node
      return
    end

    temp = @head

    @head = node
    @head.next = temp
    temp = nil
  end

  # This method removes and returns the first node in the Linked List and must set Linked List's head to the second node.
  def remove_front
    if !@head
      return
    end

    temp = @head

    @tail = @head == @tail ? nil : @tail
    @head = @head.next
    temp = nil
  end
end